WordPress/gutenberg
# Block Bindings: Disable editing of bound block attributes in editor UI

## What?
As discussed in this Slack [thread](https://wordpress.slack.com/archives/C02QB2JS7/p1705917394175109), I'm splitting [this other pull request](https://github.com/WordPress/gutenberg/pull/58059) to handle first just the use case where the editor understands the bindings attribute and adapt the UI accordingly.

As part of that, in this PR I'm doing:
* Provide an action to register a new source in the editor: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-4d145a0c43689f5a56c00a8e2919550768ebd233e91a6c6d04fee71a44bae9f2R364-R372).
* Provide some selectors to get all the registered sources or a specific one: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-23e851c3077ed95d23bb1669da8bafbae3f1e2e952897987e45b49555004dd83R289-R296).
* Add filter to the `BlockEdit` to modify `attributes` and `setAttributes` and make it understand the bindings: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R31-R100).
* Use addFilter to add the needed context to all the allowed blocks: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R126-R144).
* Adapt the paragraph, heading, button, and image blocks UI to identify when bindings exist.
* Register the post meta source using these mechanisms: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-cf4447b4134dbf24fa2819703999f0131449fb79a0611bd544c5e05fd80aa36bR21-R47).

## Why?
Splitting the original pull request should make it easier to review and it could help us land something sooner that covers understanding the bindings from the editor, although it doesn't include any UI to create/modify the bindings.

## Testing Instructions

For all the testing, we have to go to Gutenberg-Experiments and enable the Test Block Bindings option.

0. Register a new custom field however you prefer. You can use a snippet similar to this:

```PHP
register_meta(
	'post',
	'text_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'This is the content of the text custom field',
	)
);
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

### In a page

**_Test paragraph_**
1. Add a paragraph with the content connected to a custom field:

```

<p>Content of the PAGE text custom field</p>

```

2. Add a paragraph without any bindings.
3. Check that the paragraph with the bindings became non-editable.
4. Checkthat the paragraph shows the content of the custom field.
5. Check that the normal paragraph works as expected.
6. Go to the frontend and check that it works as well there.

**_Test heading_**

Repeat the paragraph test but using a heading.

**_Test button_**

1. Add a button with the text connected to a custom field.
2. Add another button with the URL connected to a custom field.
3. Add a button with the text and the URL connected to custom fields.

```

<div class="wp-block-buttons">
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">ANY TEXT</a></div>



<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">URL</a></div>



<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">BOTH</a></div>
</div>

```

4. Check that for the buttons with the text connected, they are not editable and they show the content of the custom field.
5. Check that for the buttons with the URL connected, the buttons to change the URL disappear from the UI.
6. Check that everything works in the frontend.

**_Test image_**

1. Add an image with the URL connected to a custom field.
2. Add an image with the alt attribute connected to a custom field.
3. Add an image with the title attribute connected to a custom field.
4. Add an image with more than one connection.

```

<h3 class="wp-block-heading">Image url and alt connected</h3>



<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/03/3bhkrj58Vtu7enYsRolD1fZdja1-683x1024.jpg" alt="Content of the PAGE text custom field" class="wp-image-135" title="Image title"/></figure>



<h3 class="wp-block-heading">Alt connected</h3>



<figure class="wp-block-image size-large"><img src="http://custom-fields.local/wp-content/uploads/2023/11/1-1024x1024.jpg" alt="Content of the PAGE text custom field" class="wp-image-133" title="Image title"/></figure>

```

5. For the images with the URL connected, check that the related buttons to link an image have disappeared and the image shows the URL from the custom field.
6. For the images with the alt/title connected, check that, in the right sidebar, they are disabled. Showing the value of the custom field instead and a message saying it is connected to custom fields.

### In a template

Go to a page template, for example, and repeat the process. In this case, the blocks can't show the value of the custom fields because it depends on each page. They should show a placeholder instead.

## Remaining issues
I'll try to gather here the remaining issues related to this pull request. So far I'd like to work on:
* Fix the query loop not showing the content properly.
* Review the whole UX.

## Screenshots or screencast

https://github.com/WordPress/gutenberg/assets/34552881/dccbf5f1-2148-4297-8a63-f7dc6fb5e043
